### PR TITLE
[RFC] vim-patch:7.4.{835,843,877}

### PR DESCRIPTION
**Pull request by jbradaric**
```
vim-patch:7.4.835

Problem:    Comparing utf-8 sequences does not handle different byte sizes
            correctly.
Solution:   Get the byte size of each character. (Dominique Pelle)
```

https://github.com/vim/vim/commit/f6470c288cb6f8efd60a507baf2c070f9d209ae6

```
vim-patch:7.4.843

Problem:    Still possible to go beyond the end of a string.
Solution:   Check for NUL also in second string. (Dominique Pelle)
```

https://github.com/vim/vim/commit/d43f0951bca162d4491d57df9277b5dbc462944f

```
vim-patch:7.4.877

Problem:    ":find" sometimes fails. (Excanoe)
Solution:   Compare current characters instead of previous ones.
```

https://github.com/vim/vim/commit/4d0c7bc74ac6fad5cb599dc3ade6996e848d83b6

### CODE REVIEW

File: src/nvim/file_search.c [hunk 1]
```diff
@@ -1067,18 +1067,19 @@ static int ff_wc_equal(char_u *s1, char_u *s2)
   if (s1 == NULL || s2 == NULL)
     return FALSE;
 
-  if (STRLEN(s1) != STRLEN(s2))
-    return FAIL;
-
-  for (i = 0; s1[i] != NUL && s2[i] != NUL; i += MB_PTR2LEN(s1 + i)) {
+  for (i = 0, j = 0; s1[i] != NUL;) {
```
Comment (fwalch) [inline src/nvim/file_search.c]: You could pull the variable definitions for i and j into here.

File: src/nvim/path.c [hunk 2]
```diff
@@ -1775,19 +1775,20 @@ bool same_directory(char_u *f1, char_u *f2)
  */
 int pathcmp(const char *p, const char *q, int maxlen)
 {
-  int i;
+  int i, j;
   int c1, c2;
   const char  *s = NULL;
 
-  for (i = 0; maxlen < 0 || i < maxlen; i += MB_PTR2LEN((char_u *)p + i)) {
+  for (i = 0, j = 0; maxlen < 0 || (i < maxlen && j < maxlen);) {
```
Comment (fwalch) [inline src/nvim/path.c]: Same as above.

Comment (fwalch) [inline src/nvim/path.c]: Same as above.

**Comment by fwalch** [comment]
Apart from minor comments, LGTM!

File: src/nvim/path.c (hunk 2 above)
Comment (jbradaric) [inline src/nvim/path.c]: These cannot be moved because `i` is used outside of the loop, when checking for a trailing slash.

**Comment by jbradaric** [comment]
@fwalch I will also add patch 7.4.843 to this PR because it seems that 7.4.835 didn't fix the problem completely.

**Comment by jbradaric** [comment]
I've also added patch 7.4.877 since it applies to the same part of the code as the previous patches.

**Pull request merged**